[Task] Move sitemaps and redirects to SEO Bundle

martineiber opened the pull request:
<!--

Before working on a contribution, you must determine on which branch you need to work:
- Bug fix: choose the latest maintenance branch `10.5`
- Feature/Improvement: choose `11.x` 

> All bug fixes merged into the latest maintenance branch are also merged to the current dev branch (`11.x`) on a regular basis.

## Please make sure your PR complies with all of the following points: 
- [ ] Read and accept our [contributing guidelines](/CONTRIBUTING.md) before you submit a PR.
- [ ] Features need to be proper documented in `doc/` 
- [ ] Bugfixes need a short guide how to reproduce them -> target branch is the oldest supported maintenance branch, e.g. `10.5` (see Readme.md for the list of supported versions)
- [ ] Meet all coding standards (see PhpStan actions) 

**Don't submit a PR if it doesn't comply, it'll be closed without a comment!**
-->  
  

## Changes in this pull request  
Resolves #14260

## Additional info  
Follow-up PR https://github.com/pimcore/demo/pull/417

martineiber committed:
extracted sitemap functionality to SEOBundle.

martineiber committed:
Extracted redirects functionality to SEOBundle.

martineiber committed:
Fixed PHPDoc type hint.

mattamon commented on bundles/SeoBundle/config/services.yaml:
```diff
@@ -27,4 +27,65 @@ services:
+    # REDIRECTS Export
+    #
+
+    Pimcore\Bundle\SeoBundle\Redirect\Csv: ~
```
I would move this to the bottom of the file, so all the sitemap services are in one line.

mattamon commented on bundles/SeoBundle/src/EventListener/SitemapGeneratorListener.php:
```diff
@@ -15,25 +15,25 @@
-     * @param \Iterator|GeneratorInterface[] $generators
+     * @param \IteratorAggregate|GeneratorInterface[] $generators
      *
      * TODO type hint against iterable after dropping PHP 7.0 support
```
I think we can remove this line too, since we are already at php 8.2? :)

mattamon commented on bundles/SeoBundle/src/EventListener/SitemapGeneratorListener.php:
```diff
@@ -15,25 +15,25 @@
      * TODO type hint against iterable after dropping PHP 7.0 support
      */
-    public function __construct(array|\Iterator $generators)
+    public function __construct(array|\IteratorAggregate $generators)
```
Why changing it from Iterator to IteratorAggregate?

martineiber replied:
`ServiceCollection` in `Bundles/SeoBundle/src/DependencyInjection/PimcoreSeoExtension.php`,  implements `IteratorAggregate`. That's why I changed it.

mattamon commented on bundles/SeoBundle/src/PimcoreSeoBundle.php:
```diff
@@ -52,4 +56,9 @@ public function getPath(): string
+
+    public static function registerDependentBundles(BundleCollection $collection): void
+    {
+        $collection->addBundle(PrestaSitemapBundle::class);
```
The presta/sitemap-bundle is only required within the bundle. The bundle stays in the core, so the dependency presta/sitemap-bundle stays in the core composer.json
Would be interesting if there is a way to only require this when the bundle is installed.

mattamon commented on models/DataObject/ClassDefinition/Data/UrlSlug.php:
```diff
@@ -25,7 +25,7 @@
 use Pimcore\Model\DataObject\Fieldcollection\Data\AbstractData;
 use Pimcore\Model\DataObject\Localizedfield;
-use Pimcore\Model\Redirect;
+use Pimcore\Bundle\SeoBundle\Model\Redirect;
```
What happens if the bundle ins not installed and there are slugs? 
This will break the save function right? 

Maybe we can move this to a save listener instead?

mattamon replied:
What about this one? :)

mattamon commented on lib/Maintenance/Tasks/RedirectCleanupTask.php:
```diff
@@ -17,7 +17,7 @@
 
 use Pimcore\Maintenance\TaskInterface;
-use Pimcore\Model\Redirect;
+use Pimcore\Bundle\SeoBundle\Model\Redirect;
```
We should move this task to the bundle too since there should not be a dependency in the core and the maintenance. No Cleanup is needed if the bundle is not enabled or installed

mattamon replied:
We still need to move this file to the bundle.

mattamon commented on models/Document/Hardlink.php:
```diff
@@ -18,7 +18,7 @@
 use Pimcore\Model;
 use Pimcore\Model\Document;
-use Pimcore\Model\Redirect;
+use Pimcore\Bundle\SeoBundle\Model\Redirect;
```
The doDelete will break if the bundle is not installed. 
Maybe we can also add a listener for this?

mattamon replied:
This will still break. :)

mattamon commented on models/Document/Page.php:
```diff
@@ -17,7 +17,7 @@
 
 use Pimcore\Messenger\GeneratePagePreviewMessage;
-use Pimcore\Model\Redirect;
+use Pimcore\Bundle\SeoBundle\Model\Redirect;
```
Same here: The doDelete will break if the bundle is not installed.
Maybe we can also add a listener for this?

mattamon reviewed (changes requested):
The redirects table should also be installed in the Bundle Installer and not from the Installerbundle

The redirects do not open anymore in the backend.
bundles/SeoBundle/public/js/redirects.js -> You have to also change the namespaces like with the rest of the seo bundle. 

bundles/SeoBundle/src/Controller/RedirectsController.php also here the names of the routes need to be changed . 

Check the https://github.com/pimcore/product-management/issues/13#RefactoringTips for more info. 

You also need to remove the redirect menu from the toolbar.js to the startup.js in the bundle.
If the icon is not used anywhere else you can move it to the icons.css in the bundle too. 
Also the handler this.editRedirects need to be moved into the startup.js
```js
  if (user.isAllowed("redirects") && perspectiveCfg.inToolbar("extras.redirects")) {
       extrasItems.push({
           text: t("redirects"),
           iconCls: "pimcore_nav_icon_redirects",
           itemId: 'pimcore_menu_extras_redirects',
           handler: this.editRedirects
       });
   }
```

martineiber committed:
Extract redirect functionality from core to SEOBundle

* Move Redirect status_codes to SEOBundle.
* Moved Events to SEOBundle
  * Hardlink doDelete
  * Page doDelete
  * URLSlag Change
Moved redirect table install to SEOBundle.
Moved redirect Cleanup task to SEOBundle.
Rename route name.
Moved menu creation to SEOBundle.

martineiber committed:
Merge branch '11.x' into 14260-refactor-sitemaps-and-redirects-bundle

# Conflicts:
#	bundles/CoreBundle/src/DependencyInjection/Configuration.php
#	bundles/CoreBundle/src/DependencyInjection/PimcoreCoreExtension.php
#	bundles/SeoBundle/config/services.yaml
#	models/Document/Page.php

martineiber committed:
Add RecursionBlockingEventDispatchHelperTrait Trait

martineiber committed:
Add CREATE TABLE IF NOT EXISTS

martineiber committed:
Install SeoBundle with in unit tests

martineiber committed:
Install SEOBundle with in Test Suit.

martineiber committed:
Merge branch '11.x' into 14260-refactor-sitemaps-and-redirects-bundle

mattamon commented on bundles/SeoBundle/config/maintenance.yaml:
```diff
@@ -0,0 +1,16 @@
+    #
+    # Configure all maintenance tasks
+
+    Pimcore\Bundle\SeoBundle\Maintenance\RedirectCleanupTask:
```
Can we please add the HTTP Cleanup task here too, to be more consistent? 
Should be in the services.yaml atm.

mattamon commented on bundles/SeoBundle/public/js/startup.js:
```diff
@@ -74,6 +74,17 @@ pimcore.bundle.seo.startup = Class.create({
+            menu.extras.items.push({
+                text: t("redirects"),
+                iconCls: "pimcore_nav_icon_redirects",
+                priority: 1,
```
We should maybe leave some space for other bundles, how about we set it to 5 to avoid conflicts.

mattamon commented on bundles/SeoBundle/src/EventListener/SitemapGeneratorListener.php:
```diff
@@ -15,25 +15,24 @@
     /**
-     * @param \Iterator|GeneratorInterface[] $generators
      *
      * TODO type hint against iterable after dropping PHP 7.0 support
```
```suggestion
```

mattamon commented on bundles/SeoBundle/config/services.yaml:
```diff
@@ -28,15 +28,66 @@ services:
     #
 
     Pimcore\Bundle\SeoBundle\EventListener\ResponseExceptionListener: ~
+    Pimcore\Bundle\SeoBundle\EventListener\UrlSlugUpdateListener: ~
```
Maybe we should also split this up into it's own file?
Wdyt?

mattamon commented on lib/Event/UrlSlugEvents.php:
```diff
@@ -0,0 +1,29 @@
+     *
+     * @var string
+     */
+    const POST_SAVE = 'pimcore.model.oataobject.classdefinition.urlslug.postsave';
```
```suggestion
    const POST_SAVE = 'pimcore.model.dataobject.classdefinition.urlslug.postsave';
```

mattamon replied:
We should still change this to dataobject instead of oataobject :)

mattamon commented on models/Document/Hardlink.php:
```diff
@@ -201,15 +200,6 @@ public function hasChildren(?bool $includingUnpublished = null): bool
-            $redirect->delete();
-        }
-
         parent::doDelete();
```
Since only the parent method is called, we can safely remove this.

mattamon commented on models/Document/Page.php:
```diff
@@ -66,15 +65,6 @@ class Page extends PageSnippet
-            $redirect->delete();
-        }
-
         parent::doDelete();
```
Here too, we can remove the doDelete.

mattamon reviewed (changes requested):
Overall looks fine to me, just some minor changes. 

Please check the https://github.com/pimcore/pimcore/blob/11.x/bundles/AdminBundle/public/js/pimcore/document/tree.js#L316, if nodes are moved we ask if redirects should be created.

In the DocumentsController are still some references to the old Redirects.
https://github.com/pimcore/pimcore/blob/11.x/bundles/AdminBundle/src/Controller/Admin/Document/DocumentController.php#L563

martineiber committed:
Removed unused deDelete().

martineiber committed:
Extract event listener and maintenance service to own config file.

martineiber committed:
Remove PHPDoc block.

martineiber committed:
Fix typo and priority.

martineiber committed:
Extract redirect generation on document move and save to SEOBundle.

martineiber committed:
Merge branch '11.x' into 14260-refactor-sitemaps-and-redirects-bundle

# Conflicts:
#	bundles/CoreBundle/src/DependencyInjection/PimcoreCoreExtension.php

martineiber committed:
Add return type to firePostMoveEvent.

mattamon commented on bundles/AdminBundle/public/js/pimcore/document/tree.js:
```diff
@@ -267,7 +267,6 @@ pimcore.document.tree = Class.create({
                     } else {
                         delete node.data.cls;
                     }
-                    pimcore.elementservice.nodeMoved("document", oldParent, newParent);
```
https://github.com/pimcore/pimcore/commit/9fccc1fb2295060a0bde7f29c3355e352a44cd14
Seems like merge was the cause of this line reappearing again.

martineiber replied:
merged 11.x into this branch.. this is solved.

mattamon commented on bundles/SeoBundle/config/pimcore/default.yaml:
```diff
@@ -15,4 +15,28 @@ pimcore_admin:
+            302: Found
+            303: See Other
+            307: Temporary Redirect
+        create_redirect_on_document_move: true
```
I am not sure if the standard should be true since then there is a lot going on, without the user noticing it, since we do not have a warning anymore. Maybe we set it to false by default?

mattamon commented on bundles/SeoBundle/config/services.yaml:
```diff
@@ -24,19 +24,58 @@ services:
+    # Services valid for all elements
+
+    # Generic URL generator accepting a path
+    Pimcore\Bundle\SeoBundle\Sitemap\UrlGeneratorInterface: '@Pimcore\Bundle\SeoBundle\Sitemap\UrlGenerator'
```
Maybe also an own file for the sitemaps like sitemaps.yaml?

mattamon commented on bundles/SeoBundle/src/EventListener/DocumentListener.php:
```diff
@@ -0,0 +1,156 @@
+    {
+        $pimcore_seo_redirects = Pimcore::getContainer()->getParameter('pimcore_seo.redirects');
+        if ($document instanceof Document\Page || $document instanceof Document\Hardlink) {
+            if (Pimcore\Tool\Admin::getCurrentUser()->isAllowed('redirects') && $pimcore_seo_redirects['create_redirect_on_document_move']) {
```
Maybe we should add a flag in general if we want to create redirects, not only for "move"? 
Otherwise it seems a little inconsistent don't you think?

martineiber replied:
Done!

mattamon reviewed (changes requested):
Looks fine for me, just a few small concerns that need to be considered.

martineiber committed:
Set default create_redirect_on_document_move to false.

martineiber committed:
Extract services definition into separate files for sitemap and redirect.

martineiber committed:
Merge branch '11.x' into 14260-refactor-sitemaps-and-redirects-bundle

martineiber committed:
Add create_redirect to all places where a redirect is created automatically.

mattamon commented on bundles/SeoBundle/config/pimcore/default.yaml:
```diff
@@ -32,7 +32,7 @@ pimcore_seo:
             303: See Other
             307: Temporary Redirect
-        create_redirect_on_document_move: false
+        create_redirect: false
```
I know this is nitpicky bit could we name it create_redirects plural? :)

mattamon commented on doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md:
```diff
@@ -200,11 +200,11 @@ pimcore:
 - [Authentication] Deprecated method `Pimcore\Tool\Authentication::authenticatePlaintext()` has been removed.
 - [DataObjects][CSV Export] Changed encoding of table data-types to `json_encode` from `base64_encoded`.
-- [Documents] The Message-Box whether a redirect should be created when a document is moved in the tree has been removed. This can now be set with the following configuration:
+- [Documents] The Message-Box whether a redirect should be created when a document is moved in the tree has been removed. This can now be set with the following configuration. This also counts for URLSlugs and pretty URLs
```
```suggestion
- [Documents] Moving a document in the tree no longer opens the redirect prompt asking to create redirects. Creating a redirect is now configurable with pimcore:redirects:create_redirects. This config includes URLSlugs and Pretty URLs.
```

martineiber committed:
Minor text changes.

mattamon reviewed (dismissed):
I uninstalled the bundle, but it is still enabled in the bundles.php
Unfortunately I get the following error now.

![image](https://user-images.githubusercontent.com/38959016/221128777-e7386a3a-3512-4479-a4bd-5a64b3f4de3a.png)

Can you please check this?

dvesh3 commented on bundles/AdminBundle/public/js/pimcore/document/tree.js:
```diff
@@ -300,27 +300,7 @@ pimcore.document.tree = Class.create({
-            (node.data.type === 'page' || node.data.type === 'hardlink') &&
-            pimcore.globalmanager.get("user").isAllowed('redirects')
-        ) {
-            this.nodesToMove.push({
```
should we also get rid of nodesToMove property?

martineiber replied:
Removed.

dvesh3 commented on bundles/CoreBundle/src/EventListener/Frontend/RoutingListener.php:
```diff
@@ -88,13 +84,6 @@ public function onKernelRequest(RequestEvent $event): void
-        if ($response) {
-            $event->setResponse($response);
-
-            return;
```
the control is not returned anymore that means it get executed even RedirectResponse exists from SeoBundle\RoutingListener...right?

martineiber replied:
Yes, you are right. But IMHO the `$event->setResponse($response);` will not be overwritten afterwards so it makes no different.
wdyt?

dvesh3 commented on doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md:
```diff
@@ -200,6 +200,12 @@ pimcore:
 - [Authentication] Deprecated method `Pimcore\Tool\Authentication::authenticateHttpBasic()` has been removed.
 - [Authentication] Deprecated method `Pimcore\Tool\Authentication::authenticatePlaintext()` has been removed.
 - [DataObjects][CSV Export] Changed encoding of table data-types to `json_encode` from `base64_encoded`.
+- [Documents] Moving a document in the tree no longer opens the redirect prompt asking to create redirects. Creating a redirect is now configurable with `pimcore:redirects:create_redirects`. This config includes URLSlugs and Pretty URLs.
```
Also adapt SEO Bundle upgrade notes that Redirects & Sitemap moved into the bundle

martineiber replied:
Done.

dvesh3 reviewed (changes requested):
Also update docs:
- https://pimcore.com/docs/pimcore/current/Development_Documentation/MVC/Routing_and_URLs/Redirects.html

- https://pimcore.com/docs/pimcore/current/Development_Documentation/Tools_and_Features/Sitemaps.html

that these features require `PimcoreSeoBundle` to be enabled and installed.

dvesh3 commented on bundles/SeoBundle/src/DependencyInjection/Configuration.php:
```diff
@@ -0,0 +1,99 @@
+                                ->prototype('scalar')
+                            ->end()
+                        ->end()
+                        ->booleanNode('create_redirects')
```
Should we rename the config to make it self explanatory e.g. `create_redirects_on_document_move` ? this should be documented on Redirects page

mattamon replied:
But it is not only used for documents_move. It is a general flag if you want to create redirects. As I mentioned in the daily, just because you install the SeoBundle, does not mean, that you automatically want redirects to be created.

dvesh3 replied:
Still this is quite generic and hard to understand. It could be something like `auto_create_redirects` and we should also add description to the config node.

martineiber committed:
Add guarding statement to Event listener if bundle is installed.

martineiber committed:
Remove unused nodesToMove property.

martineiber committed:
Add Documentation.

dvesh3 committed:
Docs optimizations

dvesh3 committed:
Refactor create_redirects config

dvesh3 reviewed (changes requested):
One last thing: right now renaming a document has a [checkbox ](https://github.com/pimcore/pimcore/blob/de7cbeee6cdb4e272f5529893967df30563488ef/bundles/AdminBundle/public/js/pimcore/elementservice.js#L653) to create a redirect. The logic should be moved to the SeoBundle.
![image](https://user-images.githubusercontent.com/5137917/221207567-30fbf99c-7eba-4608-b7c4-f8de7e013f8f.png)

martineiber committed:
Removed create_redirects checkbox on rename dialog.

mattamon commented on doc/Development_Documentation/02_MVC/04_Routing_and_URLs/04_Redirects.md:
```diff
@@ -1,5 +1,7 @@
 # URLs based on Redirects
 
+To use this feature please enable the `PimcoreSeoBundle` in your bundles.php file.
```
Maybe we should also mentioned that you have to install the bundle. Just enabling it, won't work right?

mattamon commented on doc/Development_Documentation/18_Tools_and_Features/39_Sitemaps.md:
```diff
@@ -1,4 +1,5 @@
 # Sitemaps
+To use this feature please enable the `PimcoreSeoBundle` in your bundles.php file.
```
Same here, with the install.

martineiber committed:
Update Documentation.

mattamon commented:
@martineiber can you please merge the latest changes and then I think we are good to merge. :)

martineiber committed:
Merge branch '11.x' into 14260-refactor-sitemaps-and-redirects-bundle

# Conflicts:
#	bundles/CoreBundle/config/pimcore/default.yaml

dvesh3 commented on doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md:
````diff
@@ -205,6 +205,12 @@ pimcore:
+```yaml
+pimcore_seo:
+    redirects:
+        create_redirect: true
````
```suggestion
        auto_create_redirect: true
```

dvesh3 committed:
Update doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md

dvesh3 commented on doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md:
````diff
@@ -205,6 +205,12 @@ pimcore:
+```yaml
+pimcore_seo:
+    redirects:
+        auto_create_redirect: true
````
```suggestion
        auto_create_redirects: true
```

dvesh3 committed:
Update doc/Development_Documentation/23_Installation_and_Upgrade/09_Upgrade_Notes/README.md
